content/concepts/hole-punching.md

```diff
@@ -0,0 +1,193 @@
+
+In the case where the other options aren't sufficient, networks can 
+use a technique called hole punching to establish connections with 
+non-public nodes.
+
+Each node connects to an unrestricted third-party server and 
+shares its external and internal address and port information.
+The server temporarily stores the node's information and relays 
+each node's information to the other. Clients can use this 
+information to establish direct connections with each other.
+
+Take two nodes, `A` and `B`, that would like the dial each other:
+
+1. The first packet of both nodes (e.g., in the case of TCP, an SYN) 
+   passes through their respective routers.
+2. The routers add a 5-tuple to their router's state table. 
+3. `Packet A` and `B` "punch holes" into their respective routers' 
+   firewalls.
+4. Both packets arrive at the opposite router.
+5. Once `A`'s packet arrives at `router B`, `router B` checks its state 
+   table and finds a 5-tuple previously added through the packet sent by 
+   node B. 
+6. The routers forward the packets through the "punched holes". 6. It 
+   then forwards the packet to `B`. The same occurs with `B`'s packet; 
+   upon arriving at `router A`, it matches a 5-tuple in `router A`'s state 
+   table and thus forwards the packet to `A`.
+  
+The following use case diagram illustrates the above process.
+
+![](https://i.imgur.com/0k2Zlj3.png)
+
+{{% notice "note" %}}
+This process assumes a mechanism to synchronize `A` and `B` simultaneously.
+{{% /notice %}}
+
+## Hole punching in libp2p
+
+Inspired by the 
+[ICE protocol](https://datatracker.ietf.org/doc/html/rfc8445) 
+specified by the IETF, libp2p includes a decentralized hole punching 
+feature that allows for firewall and NAT traversal without the need 
+for central coordination servers like STUN and TURN. 
+
+The following sequence diagram illustrates the whole process.
+
+![](https://i.imgur.com/sMGMfGZ.png)
+
+libp2p hole punching can be derived in two phases, a preparation phase and 
+a hole punching phase.
+
+### Phase I: Preparation
+
+1. [AutoNAT](/concepts/nat/#autonat): Determine whether a node is dialable, 
+   as in, discover if a node is behind a NAT or firewall.
+   
+   > This is equivalent to the 
+   > [STUN protocol](https://www.rfc-editor.org/rfc/rfc3489) in ICE.
+
+   ![](https://i.imgur.com/NeufUm7.png)
+   
+   - `B` reaches out to `Other_Peers` (e.g., boot nodes) on the network it 
+     is on and asks each node to dial it on a set of addresses it suspects 
+     could be reachable. 
+   - `Other_Peers`attempt to dial each of `B`'s addresses and report the 
+     outcome back to `B`. 
+   - Based on the reports, `B` can gauge whether it is publicly dialable and 
+     determine if hole punching is needed.
+   
+<!-- to add routing reference when available -->
+1. Routing: Discover the k-closest public Relay nodes using a lookup method 
+   (e.g. IPFS uses Kademlia DHT): `/<RELAY_ADDR>/p2p-circuit/<PEER_ID_B>`
```

Comment:
That’s an implementation detail of kubo. libp2p’s AutoRelay is happy to take relay candidates from any source.